File: 0387-first-unique-character-in-a-string/0387-first-unique-character-in-a-string.cpp
class Solution {
public:
    int firstUniqChar(string s) {

        unordered_map<char,int> umap;

        // Pass 1: count frequency
        for(char c : s){
            umap[c]++;
        }

        // Pass 2: find first unique character
        for(int i = 0; i < s.length(); i++){
            if(umap[s[i]] == 1){
                return i;
            }
        }

        return -1;
    }
};

File: 0387-first-unique-character-in-a-string/0387-first-unique-character-in-a-string.py
class Solution:
    # 5. First Unique Character in a String (LeetCode 387)
    def firstUniqChar(self, s: str) -> int:
        # Count the frequency of each character
        count = {}
        for char in s:
            count[char] = count.get(char, 0) + 1
            
        # Find the index of the first character with a frequency of 1
        for i, char in enumerate(s):
            if count[char] == 1:
                return i
        return -1